In [3]:
import pandas as pd

In [5]:
obj = pd.Series([4,7,-5,3])

obj

0    4
1    7
2   -5
3    3
dtype: int64

In [8]:
obj.values
print(obj.index)

RangeIndex(start=0, stop=4, step=1)


In [10]:
obj2 = pd.Series([4,7,-5,3], index = ['d','b','a','c'])
obj2.index

Index(['d', 'b', 'a', 'c'], dtype='object')

In [13]:
obj2[['a','d','c']]

a   -5
d    4
c    3
dtype: int64

In [15]:
'a' in obj2

True

In [28]:
sdata = {'Lagos': 35000, 'Texas': 71000, 'Oregon': 16000, 'Utah': 5000}
obj3 = pd.Series(sdata)
obj3

Lagos      35000
Texas     71000
Oregon    16000
Utah       5000
dtype: int64

In [29]:
states = ['California', 'Lagos', 'Oregon', 'Texas', 'Utah']
obj4 = pd.Series(sdata, index=states)
obj4


California        NaN
Lagos          35000.0
Oregon        16000.0
Texas         71000.0
Utah           5000.0
dtype: float64

In [34]:
obj4['Lagos'] = 65000

obj4

California        NaN
Lagos          65000.0
Oregon        16000.0
Texas         71000.0
Utah           5000.0
dtype: float64

In [27]:
# pd.isnull(obj3)

pd.notnull(obj3)


California    False
Lagos           True
Oregon         True
Texas          True
Utah           True
dtype: bool

In [25]:
#to find the NaN positions, we could use pd.isnull() and pd.notnull()

obj3.isnull() #this is a series instance method
# obj3.notnull()  

California     True
Lagos          False
Oregon        False
Texas         False
Utah          False
dtype: bool

Operations on Series
* Addition
* Subtraction
* Division

In [35]:
# Addition
obj3 + obj4

California         NaN
Lagos          100000.0
Oregon         32000.0
Texas         142000.0
Utah           10000.0
dtype: float64

In [36]:
# Division
obj3/obj4

California         NaN
Lagos          0.538462
Oregon        1.000000
Texas         1.000000
Utah          1.000000
dtype: float64

In [37]:
# integer division
obj3//obj4

California    NaN
Lagos          0.0
Oregon        1.0
Texas         1.0
Utah          1.0
dtype: float64

In [41]:
# Both the Series object itself and its index have a name attribute,
obj4.name = 'population'
obj4.index.name = 'state'

obj4

state
California        NaN
Lagos          65000.0
Oregon        16000.0
Texas         71000.0
Utah           5000.0
Name: population, dtype: float64

DataFrame

A DataFrame represents a tabular, spreadsheet-like data structure containing an ordered
collection of columns, each of which can be a different value type (numeric,
string, boolean, etc.). The DataFrame has both a row and column index; it can be
thought of as a dict of Series (one for all sharing the same index)

In [68]:
data = {'state': ['Lagos', 'Lagos', 'Lagos', 'Kano', 'Kano'],
        'year': [2000, 2001, 2002, 2001, 2002],
        'pop': [1.5, 1.7, 3.6, 2.4, 2.9]}
frame = pd.DataFrame(data)
frame

state  year  pop
0  Lagos  2000  1.5
1  Lagos  2001  1.7
2  Lagos  2002  3.6
3   Kano  2001  2.4
4   Kano  2002  2.9

In [69]:
pd.DataFrame(data, columns = ['year', 'state', 'pop'])

year  state  pop
0  2000  Lagos  1.5
1  2001  Lagos  1.7
2  2002  Lagos  3.6
3  2001   Kano  2.4
4  2002   Kano  2.9

In [70]:
frame2 = pd.DataFrame(data, columns=['year', 'state', 'pop', 'debt'],
                   index=['one', 'two', 'three', 'four', 'five'])
frame2

year  state  pop debt
one    2000  Lagos  1.5  NaN
two    2001  Lagos  1.7  NaN
three  2002  Lagos  3.6  NaN
four   2001   Kano  2.4  NaN
five   2002   Kano  2.9  NaN

In [71]:
frame2.columns

Index(['year', 'state', 'pop', 'debt'], dtype='object')

In [72]:
# A column in a DataFrame can be retrieved as a Series either by dict-like notation or by
# attribute:
frame2.state
# print(frame2['year'])
# print(frame2.year)


one      Lagos
two      Lagos
three    Lagos
four      Kano
five      Kano
Name: state, dtype: object

In [73]:
# Columns can be modified by assignment. For example, the empty 'debt' column could
# be assigned a scalar value or an array of values:
import numpy as np
frame2['debt'] = np.arange(5)
frame2


year  state  pop  debt
one    2000  Lagos  1.5     0
two    2001  Lagos  1.7     1
three  2002  Lagos  3.6     2
four   2001   Kano  2.4     3
five   2002   Kano  2.9     4

In [74]:
frame2['debt'] = pd.Series([900,210,322], index = ['two','three', 'four'])
frame2

year  state  pop   debt
one    2000  Lagos  1.5    NaN
two    2001  Lagos  1.7  900.0
three  2002  Lagos  3.6  210.0
four   2001   Kano  2.4  322.0
five   2002   Kano  2.9    NaN

In [75]:
frame2 = pd.DataFrame(data, index =([1,2,3,4,5]))
frame2

state  year  pop
1  Lagos  2000  1.5
2  Lagos  2001  1.7
3  Lagos  2002  3.6
4   Kano  2001  2.4
5   Kano  2002  2.9

In [76]:
# Assigning a column that doesn’t exist will create a new column. The del keyword will delete columns as with a dict:
frame2['western'] = frame2.state =='Lagos'
frame2

state  year  pop  western
1  Lagos  2000  1.5     True
2  Lagos  2001  1.7     True
3  Lagos  2002  3.6     True
4   Kano  2001  2.4    False
5   Kano  2002  2.9    False

In [77]:
del frame2['western']

frame2.columns

Index(['state', 'year', 'pop'], dtype='object')

In [81]:
pop = {'Kano' : {2001: 2.4, 2002 : 2.9}, 'Lagos' : 
{2000 : 1.5, 2001: 1.7, 2002 : 3.6}}

frame3 = pd.DataFrame(pop, index = [2000,2001,2002])
frame3

Kano  Lagos
2000   NaN    1.5
2001   2.4    1.7
2002   2.9    3.6

In [82]:
# transpose the result:
frame3.T

2000  2001  2002
Kano    NaN   2.4   2.9
Lagos   1.5   1.7   3.6

In [85]:
frame3.index.name = 'year'; frame3.columns.name = 'States'
frame3

States  Kano  Lagos
year               
2000     NaN    1.5
2001     2.4    1.7
2002     2.9    3.6

In [90]:
frame2.values 

array([['Lagos', 2000, 1.5],
       ['Lagos', 2001, 1.7],
       ['Lagos', 2002, 3.6],
       ['Kano', 2001, 2.4],
       ['Kano', 2002, 2.9]], dtype=object)

Index Objects

pandas’s Index objects are responsible for holding the axis labels and other metadata (like the axis name or names).

In [91]:
obj = pd.Series(range(3), index=['a', 'b', 'c'])
obj


a    0
b    1
c    2
dtype: int64

In [93]:
a_index = obj.index
a_index

Index(['a', 'b', 'c'], dtype='object')

In [95]:
# Index objects are immutable and thus can’t be modified by the user:
# a_index[1] ='d'


In [96]:
# Immutability is important so that Index objects can be safely shared among data structures:
b_index = pd.Index(np.arange(3))

obj2 = pd.Series([1.5,-2.5,0], index = b_index)

obj2

0    1.5
1   -2.5
2    0.0
dtype: float64

In [97]:
obj2.index is b_index

True

Reindexing

A critical method on pandas objects is reindex, which means to create a new object with the data conformed to a new index.

In [100]:
obj = pd.Series([4.5, 7.2, -5.3, 3.6], index=['d', 'b', 'a', 'c'])
obj

d    4.5
b    7.2
a   -5.3
c    3.6
dtype: float64

In [101]:
obj2 = obj.reindex(['a', 'b', 'c', 'd', 'e'])
obj2

a   -5.3
b    7.2
c    3.6
d    4.5
e    NaN
dtype: float64

In [103]:
obj2 = obj.reindex(['a', 'b', 'c', 'd', 'e'], fill_value=0)
obj2

a   -5.3
b    7.2
c    3.6
d    4.5
e    0.0
dtype: float64

Dropping entries from an axis

the drop method will return a new object with the indicated value or values deleted from an axis:

In [105]:
obj = pd.Series(np.arange(5.), index=['a', 'b', 'c', 'd', 'e'])
obj

a    0.0
b    1.0
c    2.0
d    3.0
e    4.0
dtype: float64

In [106]:
new_obj = obj.drop('c')

new_obj

a    0.0
b    1.0
d    3.0
e    4.0
dtype: float64

In [107]:
obj.drop(['d','c'])

a    0.0
b    1.0
e    4.0
dtype: float64

In [121]:
data = pd.DataFrame(np.arange(16).reshape((4,4)), index = ['Lagos, Abuja, Niger, Jigawa'.split(',')], columns= ['1', '2', '3', '4'])
data

1   2   3   4
Lagos     0   1   2   3
 Abuja    4   5   6   7
 Niger    8   9  10  11
 Jigawa  12  13  14  15

In [122]:
data.drop(['Lagos'])

C:\Users\Magic\AppData\Local\Temp\ipykernel_532\2939136471.py:1: PerformanceWarning: dropping on a non-lexsorted multi-index without a level parameter may impact performance.
  data.drop(['Lagos'])


1   2   3   4
 Abuja    4   5   6   7
 Niger    8   9  10  11
 Jigawa  12  13  14  15

In [123]:
data

1   2   3   4
Lagos     0   1   2   3
 Abuja    4   5   6   7
 Niger    8   9  10  11
 Jigawa  12  13  14  15

In [127]:
# data.drop(columns= '2')
data.drop(['2'], axis = 1)

1   3   4
Lagos     0   2   3
 Abuja    4   6   7
 Niger    8  10  11
 Jigawa  12  14  15

Indexing, selection, and filtering

In [132]:
# Indexing
print(data['2'])
data[['2','4']]

Lagos       1
 Abuja      5
 Niger      9
 Jigawa    13
Name: 2, dtype: int32


2   4
Lagos     1   3
 Abuja    5   7
 Niger    9  11
 Jigawa  13  15

In [133]:
data[:2]

1  2  3  4
Lagos   0  1  2  3
 Abuja  4  5  6  7

In [135]:
data

1   2   3   4
Lagos     0   1   2   3
 Abuja    4   5   6   7
 Niger    8   9  10  11
 Jigawa  12  13  14  15

In [146]:
data[data > 5] = 0
data

1  2  3  4
Lagos    0  1  2  3
 Abuja   4  5  0  0
 Niger   0  0  0  0
 Jigawa  0  0  0  0

In [144]:
data[data['2'] >= 5]

1   2   3   4
 Abuja    4   5   6   7
 Niger    8   9  10  11
 Jigawa  12  13  14  15

In [7]:
import pandas as pd
survey = pd.read_csv('survey.csv')
survey.head(10)



Year Industry_aggregation_NZSIOC Industry_code_NZSIOC Industry_name_NZSIOC  \
0  2020                     Level 1                99999       All industries   
1  2020                     Level 1                99999       All industries   
2  2020                     Level 1                99999       All industries   
3  2020                     Level 1                99999       All industries   
4  2020                     Level 1                99999       All industries   
5  2020                     Level 1                99999       All industries   
6  2020                     Level 1                99999       All industries   
7  2020                     Level 1                99999       All industries   
8  2020                     Level 1                99999       All industries   
9  2020                     Level 1                99999       All industries   

                Units Variable_code  \
0  Dollars (millions)           H01   
1  Dollars (millions)           H04   
2  Dollars (millions)           H05   
3  Dollars (millions)           H07   
4  Dollars (millions)           H08   
5  Dollars (millions)           H09   
6  Dollars (millions)           H10   
7  Dollars (millions)           H11   
8  Dollars (millions)           H12   
9  Dollars (millions)           H13   

                                     Variable_name      Variable_category  \
0                                     Total income  Financial performance   
1  Sales, government funding, grants and subsidies  Financial performance   
2                Interest, dividends and donations  Financial performance   
3                             Non-operating income  Financial performance   
4                                Total expenditure  Financial performance   
5                           Interest and donations  Financial performance   
6                                   Indirect taxes  Financial performance   
7                                     Depreciation  Financial performance   
8                          Salaries and wages paid  Financial performance   
9                         Redundancy and severance  Financial performance   

     Value                             Industry_code_ANZSIC06  
0  733,258  ANZSIC06 divisions A-S (excluding classes K633...  
1  660,630  ANZSIC06 divisions A-S (excluding classes K633...  
2   54,342  ANZSIC06 divisions A-S (excluding classes K633...  
3   18,285  ANZSIC06 divisions A-S (excluding classes K633...  
4  654,872  ANZSIC06 divisions A-S (excluding classes K633...  
5   32,730  ANZSIC06 divisions A-S (excluding classes K633...  
6    7,509  ANZSIC06 divisions A-S (excluding classes K633...  
7   26,821  ANZSIC06 divisions A-S (excluding classes K633...  
8  119,387  ANZSIC06 divisions A-S (excluding classes K633...  
9      305  ANZSIC06 divisions A-S (excluding classes K633...

In [9]:
survey.shape

(37080, 10)

In [10]:
survey.dtypes

Year                            int64
Industry_aggregation_NZSIOC    object
Industry_code_NZSIOC           object
Industry_name_NZSIOC           object
Units                          object
Variable_code                  object
Variable_name                  object
Variable_category              object
Value                          object
Industry_code_ANZSIC06         object
dtype: object

In [11]:
survey.tail(3)

Year Industry_aggregation_NZSIOC Industry_code_NZSIOC  \
37077  2013                     Level 3                 ZZ11   
37078  2013                     Level 3                 ZZ11   
37079  2013                     Level 3                 ZZ11   

             Industry_name_NZSIOC       Units Variable_code  \
37077  Food product manufacturing  Percentage           H39   
37078  Food product manufacturing  Percentage           H40   
37079  Food product manufacturing  Percentage           H41   

                Variable_name Variable_category Value  \
37077        Return on equity  Financial ratios    12   
37078  Return on total assets  Financial ratios     5   
37079   Liabilities structure  Financial ratios    46   

                                  Industry_code_ANZSIC06  
37077  ANZSIC06 groups C111, C112, C113, C114, C115, ...  
37078  ANZSIC06 groups C111, C112, C113, C114, C115, ...  
37079  ANZSIC06 groups C111, C112, C113, C114, C115, ...

In [35]:
ff = pd.read_csv('https://raw.githubusercontent.com/dphi-official/Datasets/master/exam_scores.csv')


In [36]:
ff.to_csv('exams_score.csv', index=False)

In [37]:
df = pd.read_csv('exams_score.csv')

In [38]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0    male        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2    male        group C                some college  free/reduced   
3  female        group D           bachelor's degree  free/reduced   
4    male        group D                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          74             68             67  
1               completed          58             68             66  
2                    none          66             65             65  
3                    none          74             75             73  
4                    none          78             77             71

In [59]:
df['lunch'].head(4)

0        standard
1        standard
2    free/reduced
3    free/reduced
Name: lunch, dtype: object

In [67]:
# # this conditional indexing
df['gender']=='male'

0       True
1      False
2       True
3      False
4       True
       ...  
995    False
996    False
997    False
998    False
999     True
Name: gender, Length: 1000, dtype: bool

In [68]:
# this conditional indexing, Get all the columns of the row that satisfies the condition
df[df['gender']=='male'].head()

gender race/ethnicity parental level of education         lunch  \
0   male        group B           bachelor's degree      standard   
2   male        group C                some college  free/reduced   
4   male        group D                some college      standard   
7   male        group E            some high school  free/reduced   
8   male        group C            some high school      standard   

  test preparation course  math score  reading score  writing score  
0                    none          74             68             67  
2                    none          66             65             65  
4                    none          78             77             71  
7                    none          54             50             48  
8               completed          65             58             55

In [49]:
# the columns can be indexed this way
df[['gender','reading score']]

gender  reading score
0      male             68
1    female             68
2      male             65
3    female             75
4      male             77
..      ...            ...
995  female             77
996  female             81
997  female             67
998  female             68
999    male             57

[1000 rows x 2 columns]

Index-Based Selection

* Index-based selection is to select data based on its numerical position in DataFrame.
* iloc is used for selecting data based on numerical position.
* the syntax is df.iloc[ ]

In [42]:
# this selects the first row and first element in the first column
df.iloc[0,0]

'male'

In [43]:
# this selects from row0 to row2
df.iloc[0:3]

gender race/ethnicity parental level of education         lunch  \
0    male        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2    male        group C                some college  free/reduced   

  test preparation course  math score  reading score  writing score  
0                    none          74             68             67  
1               completed          58             68             66  
2                    none          66             65             65

In [44]:
# this selects from row0 to row3 and column 4
df.iloc[0:4,5]

0    74
1    58
2    66
3    74
Name: math score, dtype: int64

In [45]:
df.iloc[[0,1,2,3,4], :4]

gender race/ethnicity parental level of education         lunch
0    male        group B           bachelor's degree      standard
1  female        group C                some college      standard
2    male        group C                some college  free/reduced
3  female        group D           bachelor's degree  free/reduced
4    male        group D                some college      standard

Label-Based Selection

This function selects data by the label of the rows and columns. The df.loc indexer selects data in a different way than just the indexing operator. It can select subsets of rows or columns. It can also simultaneously select subsets of rows and columns.

In [53]:
# In order to select a single row using .loc[], we put a single row label in a .loc function.
df.loc[0]

gender                                      male
race/ethnicity                           group B
parental level of education    bachelor's degree
lunch                                   standard
test preparation course                     none
math score                                    74
reading score                                 68
writing score                                 67
Name: 0, dtype: object

In [54]:
# In order to select multiple rows, we put all the row labels in a list and pass that to .loc function.
df.loc[[0,3]]

gender race/ethnicity parental level of education         lunch  \
0    male        group B           bachelor's degree      standard   
3  female        group D           bachelor's degree  free/reduced   

  test preparation course  math score  reading score  writing score  
0                    none          74             68             67  
3                    none          74             75             73

In [56]:
# to select some rows in a columns
df.loc[0:4,'math score']

0    74
1    58
2    66
3    74
4    78
Name: math score, dtype: int64

In [57]:
# to select some rows in a some columns
df.loc[0:4,['math score', 'race/ethnicity']]

math score race/ethnicity
0          74        group B
1          58        group C
2          66        group C
3          74        group D
4          78        group D

Summary

In [95]:
df.describe()

math score  reading score  writing score
count  1000.000000     1000.00000    1000.000000
mean     67.128000       70.17400      68.973000
std      14.815367       14.85599      15.109155
min      15.000000       18.00000      10.000000
25%      58.000000       60.00000      59.000000
50%      67.000000       70.00000      69.000000
75%      78.000000       81.00000      80.000000
max     100.000000      100.00000     100.000000

In [96]:
df.describe(include= 'object')


gender race/ethnicity parental level of education     lunch  \
count     1000           1000                        1000      1000   
unique       2              5                           6         2   
top     female        group C                some college  standard   
freq       502            294                         226       649   

       test preparation course  
count                     1000  
unique                       2  
top                       none  
freq                       654

In [97]:
df.describe(include='all')


gender race/ethnicity parental level of education     lunch  \
count     1000           1000                        1000      1000   
unique       2              5                           6         2   
top     female        group C                some college  standard   
freq       502            294                         226       649   
mean       NaN            NaN                         NaN       NaN   
std        NaN            NaN                         NaN       NaN   
min        NaN            NaN                         NaN       NaN   
25%        NaN            NaN                         NaN       NaN   
50%        NaN            NaN                         NaN       NaN   
75%        NaN            NaN                         NaN       NaN   
max        NaN            NaN                         NaN       NaN   

       test preparation course   math score  reading score  writing score  
count                     1000  1000.000000     1000.00000    1000.000000  
unique                       2          NaN            NaN            NaN  
top                       none          NaN            NaN            NaN  
freq                       654          NaN            NaN            NaN  
mean                       NaN    67.128000       70.17400      68.973000  
std                        NaN    14.815367       14.85599      15.109155  
min                        NaN    15.000000       18.00000      10.000000  
25%                        NaN    58.000000       60.00000      59.000000  
50%                        NaN    67.000000       70.00000      69.000000  
75%                        NaN    78.000000       81.00000      80.000000  
max                        NaN   100.000000      100.00000     100.000000



Let's see what information about the data is returned in the above table:

    count - the count of non-null entries in the particular column. For example, gender column has 1000 non-null entries.
    unique - the count of unique values in a column. Only for categorical columns. For example, gender column has 2 unique values - male and female.
    top - this is also for only categorical columns. This tells us which category is occurring maximum number of times. For example in gender column 'female' is occurring maximum number of times.
    freq - this is again for categorical columns only. This tells you the number of occurrences of the top category in that column. For example, 'female' in gender column is occurring 502 times.
    mean - the mean value of the numerical column. For example, the mean math score is 67.128.
    std - this is the standard deviation of the numerical column. This tells you about the variation in the data. Don't worry if you don't know about it.
    min - the minimum value in the numerical column. For example, the minimum math score is 15.0
    25% - the 25th percentile (or 1st quartile) value in the numerical column. For example, the 25th percentile value for math score is 58.0.
    50% - the 50th percentile (or 2nd quartile or the median) value in the numerical column. For example, the median math score is 67.0.
    75% - the 75th percentile (or 3rd quartile) value in the numerical column. For example, the 75th percentile value for math score is 78.0.
    max - the maximum value in the numerical column. For example, the maximum math score is 100.0.
    NaN values means that for a particular column, a particular summary value is not available. For example, gender (a categorical column), does not have mean value or median value as these are the properties of a numerical column only.



In [98]:
df.gender.unique()

array(['male', 'female'], dtype=object)

In [106]:
df['math score'].mean()


67.128

Sorting

In [123]:
df.sort_values(by= 'math score').head()

gender race/ethnicity parental level of education         lunch  \
949    male        group C                 high school  free/reduced   
854    male        group C                 high school  free/reduced   
891  female        group C                 high school  free/reduced   
349    male        group B            some high school  free/reduced   
739  female        group E            some high school  free/reduced   

    test preparation course  math score  reading score  writing score  
949                    none          15             18             10  
854                    none          18             30             18  
891                    none          23             31             27  
349                    none          25             31             30  
739                    none          25             35             38

In [124]:
df.sort_values(by='math score', ascending= False).head()


gender race/ethnicity parental level of education     lunch  \
91    male        group D          associate's degree  standard   
776   male        group E          associate's degree  standard   
588   male        group D                some college  standard   
128   male        group A          associate's degree  standard   
565   male        group E           bachelor's degree  standard   

    test preparation course  math score  reading score  writing score  
91                     none         100             94             96  
776               completed         100            100            100  
588               completed         100             85             91  
128               completed         100             97             94  
565               completed         100            100            100

Checking and Filling Missing Data
* If in any row or column in a DataFrame, a value is not available, it is said to be a missing value.
* So, defining missing data: Missing data (or missing values) is defined as the data values that are not stored in a column or row.
* Pandas provides isnull(), isna() functions to detect missing values. Both of them do the same thing.

In [130]:
df.isna()

gender  race/ethnicity  parental level of education  lunch  \
0     False           False                        False  False   
1     False           False                        False  False   
2     False           False                        False  False   
3     False           False                        False  False   
4     False           False                        False  False   
..      ...             ...                          ...    ...   
995   False           False                        False  False   
996   False           False                        False  False   
997   False           False                        False  False   
998   False           False                        False  False   
999   False           False                        False  False   

     test preparation course  math score  reading score  writing score  
0                      False       False          False          False  
1                      False       False          False          False  
2                      False       False          False          False  
3                      False       False          False          False  
4                      False       False          False          False  
..                       ...         ...            ...            ...  
995                    False       False          False          False  
996                    False       False          False          False  
997                    False       False          False          False  
998                    False       False          False          False  
999                    False       False          False          False  

[1000 rows x 8 columns]

In [131]:
df.isna().sum()


gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

* Pandas also provides fillna() method to fill the missing values. fillna() provides many different strategies to fill missing values.

In [69]:
dphi = pd.read_csv('https://raw.githubusercontent.com/dphi-official/Datasets/master/Standard_Metropolitan_Areas_Data-data.csv')


In [71]:
dphi.head(11)

land_area  percent_city  percent_senior  physicians  hospital_beds  \
0        1384          78.1            12.3       25627          69678   
1        3719          43.9             9.4       13326          43292   
2        3553          37.4            10.7        9724          33731   
3        3916          29.9             8.8        6402          24167   
4        2480          31.5            10.5        8502          16751   
5        2815          23.1             6.7        7340          16941   
6        8360          46.3             8.2        4047          14347   
7        6794          60.1             6.3        4562          14333   
8        3049          19.5            12.1        4005          21149   
9        4647          31.5             9.2        3916          12815   
10       1008          16.6            10.3        4006          16704   

    graduates  work_force  income  region  crime_rate  
0        50.1      4083.9   72100       1       75.55  
1        53.9      3305.9   54542       2       56.03  
2        50.6      2066.3   33216       1       41.32  
3        52.2      1966.7   32906       2       67.38  
4        66.1      1514.5   26573       4       80.19  
5        68.3      1541.9   25663       3       58.48  
6        53.6      1321.2   18350       3       72.25  
7        51.7      1272.7   18221       3       64.88  
8        53.4       967.5   15826       1       30.51  
9        65.1      1032.2   14542       2       55.30  
10       55.9       935.5   15953       1       54.16

In [72]:
dphi.tail(2)


land_area  percent_city  percent_senior  physicians  hospital_beds  \
97        813          13.4            10.9         371           4355   
98        654          28.8             3.9         140           1296   

    graduates  work_force  income  region  crime_rate  
97       58.0        97.0    1589       1       36.33  
98       55.1        66.9    1148       3       68.76

In [76]:
sample1 = dphi.iloc[[1,3,5,7,9,13],[0,-4,-3,-2,-1]]
sample1

land_area  work_force  income  region  crime_rate
1        3719      3305.9   54542       2       56.03
3        3916      1966.7   32906       2       67.38
5        2815      1541.9   25663       3       58.48
7        6794      1272.7   18221       3       64.88
9        4647      1032.2   14542       2       55.30
13        782       915.2   12591       4       63.20

In [82]:
sample2 = [dphi['region'] == 2]
sample2

[0     False
 1      True
 2     False
 3      True
 4     False
       ...  
 94    False
 95    False
 96    False
 97    False
 98    False
 Name: region, Length: 99, dtype: bool]

In [84]:
# this counts number of occurances in a column
dphi['region'].value_counts()

3    36
2    25
1    21
4    17
Name: region, dtype: int64

In [85]:
# pandas count unique values ascending:
dphi['region'].value_counts(ascending= True)


4    17
1    21
2    25
3    36
Name: region, dtype: int64

In [94]:
# Count occurences of certain value (i.e. male) in a column (i.e., gender)
df.gender.value_counts().male

498

In [86]:
# Counting occurences as well as missing values:
dphi['region'].value_counts(dropna= False)


3    36
2    25
1    21
4    17
Name: region, dtype: int64

In [87]:
#  descriptive statistics of the column by using the describe() method:
dphi['region'].describe()


count    99.000000
mean      2.494949
std       1.013921
min       1.000000
25%       2.000000
50%       3.000000
75%       3.000000
max       4.000000
Name: region, dtype: float64

In [107]:
dphi.mean()

land_area         2615.727273
percent_city        42.518182
percent_senior       9.781818
physicians        1828.333333
hospital_beds     6345.868687
graduates           54.463636
work_force         449.366667
income            6762.505051
region               2.494949
crime_rate          55.643030
dtype: float64

In [108]:
dphi.region.unique()

array([1, 2, 4, 3], dtype=int64)

In [114]:
dphi.describe()

land_area  percent_city  percent_senior    physicians  \
count     99.000000     99.000000       99.000000     99.000000   
mean    2615.727273     42.518182        9.781818   1828.333333   
std     3045.826210     17.348277        2.524547   3192.199763   
min       47.000000     13.400000        3.900000    140.000000   
25%     1408.000000     30.100000        8.350000    459.000000   
50%     1951.000000     39.500000        9.700000    774.000000   
75%     2890.500000     52.600000       10.750000   1911.500000   
max    27293.000000    100.000000       21.800000  25627.000000   

       hospital_beds  graduates   work_force        income     region  \
count      99.000000  99.000000    99.000000     99.000000  99.000000   
mean     6345.868687  54.463636   449.366667   6762.505051   2.494949   
std      9136.202716   7.773286   610.990885  10393.349660   1.013921   
min       481.000000  30.300000    66.900000    769.000000   1.000000   
25%      2390.000000  50.250000   150.300000   2003.000000   2.000000   
50%      3472.000000  54.000000   257.200000   3510.000000   3.000000   
75%      6386.500000  58.300000   436.500000   6283.500000   3.000000   
max     69678.000000  72.800000  4083.900000  72100.000000   4.000000   

       crime_rate  
count   99.000000  
mean    55.643030  
std     13.470943  
min     23.320000  
25%     46.115000  
50%     56.060000  
75%     63.860000  
max     85.620000

In [116]:
dphi.land_area.median()

1951.0

In [119]:
dphi.work_force.value_counts()

136.9    2
131.2    1
171.4    1
163.8    1
146.4    1
        ..
401.7    1
355.4    1
356.5    1
359.1    1
66.9     1
Name: work_force, Length: 98, dtype: int64

In [121]:
dphi.region.value_counts()


3    36
2    25
1    21
4    17
Name: region, dtype: int64

In [122]:
dphi.crime_rate.mean()

55.6430303030303

In [127]:
dphi.sort_values(by= 'crime_rate', ascending= False).head()

land_area  percent_city  percent_senior  physicians  hospital_beds  \
20       9155          53.8            11.1        2280           6450   
74       1412          39.2            11.3         436           1837   
53       5966          39.5             9.6         737           1907   
4        2480          31.5            10.5        8502          16751   
67       8152          22.3             9.1         405           1254   

    graduates  work_force  income  region  crime_rate  
20       60.1       575.2    7766       4       85.62  
74       49.4       154.2    2098       4       82.68  
53       52.7       246.6    3007       4       80.94  
4        66.1      1514.5   26573       4       80.19  
67       51.7       165.6    2257       4       78.10

In [128]:
dphi.rename(columns= {'physicians': 'num_physicians'}, inplace=True)

In [129]:
dphi.head()

land_area  percent_city  percent_senior  num_physicians  hospital_beds  \
0       1384          78.1            12.3           25627          69678   
1       3719          43.9             9.4           13326          43292   
2       3553          37.4            10.7            9724          33731   
3       3916          29.9             8.8            6402          24167   
4       2480          31.5            10.5            8502          16751   

   graduates  work_force  income  region  crime_rate  
0       50.1      4083.9   72100       1       75.55  
1       53.9      3305.9   54542       2       56.03  
2       50.6      2066.3   33216       1       41.32  
3       52.2      1966.7   32906       2       67.38  
4       66.1      1514.5   26573       4       80.19

In [146]:
tit = pd.read_csv('https://raw.githubusercontent.com/dphi-official/Datasets/master/titanic_data.csv ')

tit.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [147]:
tit.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [148]:
age_mean_before = tit.Age.mean()
age_mean_before

29.69911764705882

In [149]:
# this fills the NaN spaces in Age column with the mean of the column Age
tit.fillna(tit.Age.mean(), inplace= True)

In [150]:
age_mean_after = tit.Age.mean()
age_mean_after


29.69911764705882

In [151]:
tit.Embarked.value_counts()

S                    644
C                    168
Q                     77
29.69911764705882      2
Name: Embarked, dtype: int64

In [154]:
tit.Embarked.fillna('S', inplace= True)

In [155]:
tit.Embarked.value_counts()


S                    644
C                    168
Q                     77
29.69911764705882      2
Name: Embarked, dtype: int64

In [157]:
tit.Embarked.isna().sum()

0